File: jcode-panel/jcode_panel/popup_context.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
import re
from urllib.parse import unquote, urlparse

MAX_SELECTED_TEXT_CHARS = 4000
URL_RE = re.compile(r"https?://[^\s<>'\")]+", re.IGNORECASE)
POPUP_CONTEXT_CHIP_DELETE_RE = re.compile(r"(?:\[(?:text:[^\]]+|link:[^\]]+|file:[^\]]+|\d+ files)\])\s*$", re.IGNORECASE)
POPUP_CONTEXT_CHIP_RE = re.compile(r"\[(?:text:[^\]]+|link:[^\]]+|file:[^\]]+|\d+ files)\]", re.IGNORECASE)


@dataclass
class PopupContextChip:
    tag: str
    body: str
    kind: str


def build_popup_context_chips(
    *,
    selected_text: str = "",
    file_paths: list[str] | None = None,
    app: str = "",
    window_title: str = "",
    file_path: str = "",
    line: int | None = None,
) -> list[PopupContextChip]:
    """Build editable prompt chips from the context available when popup opens.

    This intentionally returns plain ASCII tags because Gtk.Entry cursor handling is
    reliable with those, unlike embedded widgets or unicode-heavy visual chips.
    """
    chips: list[PopupContextChip] = []
    seen: set[str] = set()

    selection = _clean(selected_text)
    selection_url = _first_url(selection)
    if selection:
        if selection_url:
            _append_unique(chips, seen, _link_chip(selection_url))
        else:
            chips.append(_text_chip(selection, app=app, window_title=window_title, file_path=file_path, line=line))

    files = [str(Path(p).expanduser()) for p in (file_paths or []) if p]
    if files:
        chips.append(_files_chip(files))

    return chips


def expand_popup_context_chips(text: str, chips: list[PopupContextChip]) -> str:
    if not chips:
        return text
    remaining = text
    blocks: list[str] = []
    for chip in chips:
        if chip.tag in remaining:
            remaining = remaining.replace(chip.tag, " ")
            blocks.append(chip.body)
    prompt = re.sub(r"[ \t]{2,}", " ", remaining).strip()
    if not blocks:
        return text
    tail = "User prompt:\n" + prompt if prompt else "User prompt:\n"
    return "\n\n".join([*blocks, tail]).strip()


def strip_popup_context_chips(text: str) -> str:
    return re.sub(r"[ \t]{2,}", " ", POPUP_CONTEXT_CHIP_RE.sub("", text)).strip()


def _append_unique(chips: list[PopupContextChip], seen: set[str], chip: PopupContextChip) -> None:
    key = chip.body
    if key not in seen:
        chips.append(chip)
        seen.add(key)


def _text_chip(text: str, *, app: str, window_title: str, file_path: str, line: int | None) -> PopupContextChip:
    label = _text_label(file_path, window_title, line)
    limited, truncated = _limit_text(text, MAX_SELECTED_TEXT_CHARS)
    parts = ["Context: selected text"]
    if app:
        parts.append(f"app: {app}")
    if window_title:
        parts.append(f"window: {window_title}")
    if file_path:
        parts.append(f"file: {file_path}")
    if line:
        parts.append(f"line: {line}")
    if truncated:
        parts.append(f"note: selected text was truncated to {MAX_SELECTED_TEXT_CHARS} characters")
    parts.extend(["selected text:", limited])
    return PopupContextChip(tag=f"[text:{label}]", body="\n".join(parts), kind="text")


def _link_chip(url: str) -> PopupContextChip:
    parsed = urlparse(url)
    host = parsed.netloc or parsed.path.split("/")[0] or "link"
    label = _safe_label(host, fallback="link", max_len=36)
    return PopupContextChip(tag=f"[link:{label}]", body=f"Context: link\nurl: {url}", kind="link")


def _files_chip(paths: list[str], current_file: bool = False) -> PopupContextChip:
    clean_paths = [str(Path(p).expanduser()) for p in paths if p]
    if len(clean_paths) == 1:
        name = Path(clean_paths[0]).name or clean_paths[0]
        tag = f"[file:{_safe_label(name, fallback='file', max_len=42)}]"
        title = "Context: current file" if current_file else "Context: selected file"
        body = f"{title}\npath: {clean_paths[0]}"
        return PopupContextChip(tag=tag, body=body, kind="file")
    tag = f"[{len(clean_paths)} files]"
    body = "Context: selected files\npaths:\n" + "\n".join(f"- {p}" for p in clean_paths)
    return PopupContextChip(tag=tag, body=body, kind="file")


def _text_label(file_path: str, window_title: str, line: int | None) -> str:
    if file_path:
        base = Path(file_path).name or "selection"
        suffix = f":{line}" if line else ""
        return _safe_label(base + suffix, fallback="selection", max_len=42)
    if window_title:
        return _safe_label(window_title, fallback="selection", max_len=42)
    return "selection"


def _file_paths_from_uris(raw: str) -> list[str]:
    paths: list[str] = []
    for token in re.split(r"[\r\n]+", raw or ""):
        token = token.strip()
        if not token:
            continue
        if token.startswith("file://"):
            parsed = urlparse(token)
            paths.append(unquote(parsed.path))
        elif token.startswith("/") and Path(token).exists():
            paths.append(token)
    return list(dict.fromkeys(paths))[:50]


def _first_url(text: str) -> str:
    match = URL_RE.search(text or "")
    return match.group(0).rstrip(".,;:]") if match else ""


def _limit_text(text: str, limit: int) -> tuple[str, bool]:
    text = _clean(text)
    if len(text) <= limit:
        return text, False
    head = max(1, limit // 2)
    tail = max(1, limit - head - 80)
    return text[:head].rstrip() + "\n... [truncated] ...\n" + text[-tail:].lstrip(), True


def _clean(text: str) -> str:
    return (text or "").replace("\r\n", "\n").replace("\r", "\n").strip()


def _safe_label(text: str, *, fallback: str, max_len: int) -> str:
    label = re.sub(r"[\[\]\n\r\t]+", " ", text or "").strip()
    label = re.sub(r"\s+", " ", label)
    if len(label) > max_len:
        label = label[: max_len - 1].rstrip() + "…"
    return label or fallback

File: jcode-panel-tauri/src-tauri/src/core/popup_context.rs
use regex::Regex;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PopupContextChip {
    pub tag: String,
    pub body: String,
    pub kind: String,
}

const MAX_SELECTED_TEXT_CHARS: usize = 4000;

pub fn build_popup_context_chips(
    selected_text: &str,
    file_paths: &[String],
    app: &str,
    window_title: &str,
    file_path: &str,
    line: Option<u32>,
) -> Vec<PopupContextChip> {
    let mut chips = vec![];
    let selection = clean(selected_text);
    if !selection.is_empty() {
        if let Some(url) = first_url(&selection) {
            chips.push(link_chip(&url));
        } else {
            chips.push(text_chip(&selection, app, window_title, file_path, line));
        }
    }
    let files: Vec<String> = file_paths
        .iter()
        .filter(|p| !p.is_empty())
        .cloned()
        .collect();
    if !files.is_empty() {
        chips.push(files_chip(&files));
    }
    chips
}

pub fn expand_popup_context_chips(text: &str, chips: &[PopupContextChip]) -> String {
    let mut remaining = text.to_string();
    let mut blocks = vec![];
    for chip in chips {
        if remaining.contains(&chip.tag) {
            remaining = remaining.replace(&chip.tag, " ");
            blocks.push(chip.body.clone());
        }
    }
    if blocks.is_empty() {
        return text.into();
    }
    let prompt = Regex::new(r"[ \t]{2,}")
        .unwrap()
        .replace_all(&remaining, " ")
        .trim()
        .to_string();
    blocks.push(format!("User prompt:\n{prompt}"));
    blocks.join("\n\n").trim().to_string()
}

fn text_chip(
    text: &str,
    app: &str,
    window_title: &str,
    file_path: &str,
    line: Option<u32>,
) -> PopupContextChip {
    let label = if !file_path.is_empty() {
        safe_label(
            &format!(
                "{}{}",
                std::path::Path::new(file_path)
                    .file_name()
                    .unwrap_or_default()
                    .to_string_lossy(),
                line.map(|l| format!(":{l}")).unwrap_or_default()
            ),
            "selection",
            42,
        )
    } else if !window_title.is_empty() {
        safe_label(window_title, "selection", 42)
    } else {
        "selection".into()
    };
    let (limited, truncated) = limit_text(text, MAX_SELECTED_TEXT_CHARS);
    let mut parts = vec!["Context: selected text".to_string()];
    if !app.is_empty() {
        parts.push(format!("app: {app}"));
    }
    if !window_title.is_empty() {
        parts.push(format!("window: {window_title}"));
    }
    if !file_path.is_empty() {
        parts.push(format!("file: {file_path}"));
    }
    if let Some(line) = line {
        parts.push(format!("line: {line}"));
    }
    if truncated {
        parts.push(format!(
            "note: selected text was truncated to {MAX_SELECTED_TEXT_CHARS} characters"
        ));
    }
    parts.push("selected text:".into());
    parts.push(limited);
    PopupContextChip {
        tag: format!("[text:{label}]"),
        body: parts.join("\n"),
        kind: "text".into(),
    }
}
fn link_chip(url: &str) -> PopupContextChip {
    let host = url
        .split("//")
        .nth(1)
        .unwrap_or(url)
        .split('/')
        .next()
        .unwrap_or("link");
    PopupContextChip {
        tag: format!("[link:{}]", safe_label(host, "link", 36)),
        body: format!("Context: link\nurl: {url}"),
        kind: "link".into(),
    }
}
fn files_chip(paths: &[String]) -> PopupContextChip {
    if paths.len() == 1 {
        let name = std::path::Path::new(&paths[0])
            .file_name()
            .unwrap_or_default()
            .to_string_lossy();
        PopupContextChip {
            tag: format!("[file:{}]", safe_label(&name, "file", 42)),
            body: format!("Context: selected file\npath: {}", paths[0]),
            kind: "file".into(),
        }
    } else {
        PopupContextChip {
            tag: format!("[{} files]", paths.len()),
            body: format!(
                "Context: selected files\npaths:\n{}",
                paths
                    .iter()
                    .map(|p| format!("- {p}"))
                    .collect::<Vec<_>>()
                    .join("\n")
            ),
            kind: "file".into(),
        }
    }
}
fn first_url(text: &str) -> Option<String> {
    Regex::new(r#"https?://[^\s<>'")]+"#)
        .unwrap()
        .find(text)
        .map(|m| {
            m.as_str()
                .trim_end_matches(['.', ',', ';', ':', ']'])
                .to_string()
        })
}
fn limit_text(text: &str, limit: usize) -> (String, bool) {
    let text = clean(text);
    if text.len() <= limit {
        (text, false)
    } else {
        let head = limit / 2;
        let tail = limit - head - 80;
        (
            format!(
                "{}\n... [truncated] ...\n{}",
                text[..head].trim_end(),
                text[text.len() - tail..].trim_start()
            ),
            true,
        )
    }
}
fn clean(text: &str) -> String {
    text.replace("\r\n", "\n").replace('\r', "\n").trim().into()
}
fn safe_label(text: &str, fallback: &str, max_len: usize) -> String {
    let mut label = Regex::new(r"[\[\]\n\r\t]+")
        .unwrap()
        .replace_all(text, " ")
        .trim()
        .to_string();
    label = Regex::new(r"\s+")
        .unwrap()
        .replace_all(&label, " ")
        .to_string();
    if label.len() > max_len {
        label = format!("{}…", label[..max_len - 1].trim_end());
    }
    if label.is_empty() {
        fallback.into()
    } else {
        label
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn chips_expand_selected_text_and_urls() {
        let chips = build_popup_context_chips("hello", &[], "Code", "win", "/tmp/a.rs", Some(2));
        assert_eq!(chips[0].tag, "[text:a.rs:2]");
        assert!(
            expand_popup_context_chips("fix [text:a.rs:2]", &chips).contains("User prompt:\nfix")
        );
        let url = build_popup_context_chips("see https://example.com/a", &[], "", "", "", None);
        assert_eq!(url[0].tag, "[link:example.com]");
    }
}
